core/trino-main/src/main/java/io/trino/util/ColumnDefaultOptions.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.util;
+
+import com.google.common.primitives.Primitives;
+import io.airlift.slice.Slice;
+import io.trino.spi.expression.Constant;
+import io.trino.spi.type.CharType;
+import io.trino.spi.type.DecimalType;
+import io.trino.spi.type.Decimals;
+import io.trino.spi.type.Int128;
+import io.trino.spi.type.LongTimestamp;
+import io.trino.spi.type.LongTimestampWithTimeZone;
+import io.trino.spi.type.TimestampType;
+import io.trino.spi.type.TimestampWithTimeZoneType;
+import io.trino.spi.type.Type;
+import io.trino.spi.type.VarcharType;
+import io.trino.sql.analyzer.LiteralInterpreter;
+import io.trino.sql.tree.Literal;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.slice.SliceUtf8.countCodePoints;
+import static io.trino.spi.StandardErrorCode.INVALID_LITERAL;
+import static io.trino.spi.type.DateTimeEncoding.unpackMillisUtc;
+import static io.trino.spi.type.TimestampType.TIMESTAMP_MICROS;
+import static io.trino.spi.type.TimestampType.TIMESTAMP_PICOS;
+import static io.trino.spi.type.TimestampWithTimeZoneType.TIMESTAMP_TZ_MILLIS;
+import static io.trino.spi.type.TimestampWithTimeZoneType.TIMESTAMP_TZ_PICOS;
+import static io.trino.spi.type.Timestamps.MICROSECONDS_PER_SECOND;
+import static io.trino.sql.analyzer.SemanticExceptions.semanticException;
+import static java.lang.Math.floorMod;
+import static java.util.Locale.ENGLISH;
+
+public final class ColumnDefaultOptions
+{
+    private ColumnDefaultOptions() {}
+
+    public static Constant evaluateLiteral(LiteralInterpreter literalInterpreter, Literal literal, Type type)
+    {
+        try {
+            Object value = literalInterpreter.evaluate(literal, type);
+            checkDefaultValue(value, type);
```

Comment:
What happens if we declare column type as `double`, and pass an integer default value?
According to spec, it is legal:
```
If the subject data type is approximate numeric, then the <literal> shall be a <signed
numeric literal>.
```

In such a case, we should apply a coercion.
For the general case, we should:
- analyze the passed literal to get the type
- if the type is not equal to the declared column type, check if this combination is legal according to the spec. Most cases can probably be handled with a `TypeCoercion#canCoerce()` call.
- apply a coercion
- evaluate the expression and store as a Constant

Reply:
Answered above: https://github.com/trinodb/trino/pull/25679#discussion_r2132354936

Reply:
@ebyhr , what about ConstantEvaluator doesn't work?

We should avoid using LiteralInterpreter. That's an implementation detail of the analyzer and not meant for broader usage.

Reply:
@martint ConstantEvaluator didn't work when the declared type is smaller than the literal type. For instance, `TINYINT DEFAULT 123` throws "Cannot cast type integer to tinyint". 

https://github.com/trinodb/trino/blob/2718456e96a386d7e0ea0b6d620918061edfcda0/core/trino-main/src/main/java/io/trino/sql/analyzer/ConstantEvaluator.java#L63-L66

Reply:
Are you trying to evaluate the expression using the declared type of the column? If so, that's clearly not going to work because the evaluator doesn't understand store semantics when doing type conversions.

What you need to do instead is:
1. Analyze the expression to determine its type
2. Evaluate the expression using the type derived in 1)
3. Cast the value to the target type using the same rules used for INSERT

Reply:
Addressed 1 and 2. 

> Cast the value to the target type using the same rules used for INSERT

Is it exactly **_same_**?  The DEFAULT rule shared at https://github.com/trinodb/trino/pull/25679#discussion_r2104171709 indicates it's different from INSERT. Not only string literals but also datetime literals don't match the current INSERT behavior in my understanding.